src/clp_ffi_py/ir/native/PyDeserializer.hpp

```diff
@@ -0,0 +1,257 @@
+                clp::UtcOffset utc_offset_new
+        ) -> clp::ffi::ir_stream::IRErrorCode {
+            return m_utc_offset_change_handle(utc_offset_old, utc_offset_new);
+        }
+
+        [[nodiscard]] auto handle_schema_tree_node_insertion(
+                clp::ffi::SchemaTree::NodeLocator schema_tree_node_locator
+        ) -> clp::ffi::ir_stream::IRErrorCode {
+            return m_schema_tree_node_insertion_handle(schema_tree_node_locator);
+        }
+
+        [[nodiscard]] auto handle_end_of_stream() -> clp::ffi::ir_stream::IRErrorCode {
+            return m_end_of_stream_handle();
+        }
+
+    private:
+        // Variables
+        LogEventHandle m_log_event_handle;
+        UtcOffsetChangeHandle m_utc_offset_change_handle;
+        SchemaTreeNodeInsertionHandle m_schema_tree_node_insertion_handle;
+        EndOfStreamHandle m_end_of_stream_handle;
+    };
+
+    using Deserializer = clp::ffi::ir_stream::Deserializer<IrUnitHandler>;
+
+    // Methods
+    /**
+     * Implements `IrUnitHandler::EndOfStreamHandle`.
+     * This handle function sets the underlying `m_end_of_stream_reached` to true.
+     * @return IRErrorCode::IRErrorCode_Success on success.
+     */
+    [[maybe_unused]] auto handle_end_of_stream() -> clp::ffi::ir_stream::IRErrorCode {
+        m_end_of_stream_reached = true;
+        return clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success;
+    }
+
+    /**
+     * Implements `IrUnitHandler::LogEventHandle`.
+     * This handle function sets the underlying `m_deserialized_log_event` with the given input.
+     * @param kv_log_event
+     * @return IRErrorCode::IRErrorCode_Success on success.
+     *
+     */
+    [[nodiscard]] auto handle_log_event(clp::ffi::KeyValuePairLogEvent&& log_event
+    ) -> clp::ffi::ir_stream::IRErrorCode;
+
+    /**
+     * @return Whether `m_deserialized_log_event` has been set.
+     */
+    [[nodiscard]] auto has_unreleased_deserialized_log_event() const -> bool {
+        return nullptr != m_deserialized_log_event;
+    }
+
+    /**
+     * Releases the ownership of the underlying deserialized log event as a rvalue reference.
+     * NOTE: this method doesn't check whether the ownership is empty (nullptr). The caller must
+     * ensure the ownership is legal.
+     * @return The released ownership of the deserialized log event.
+     */
+    [[nodiscard]] auto release_deserialized_log_event() -> clp::ffi::KeyValuePairLogEvent {
+        auto released{std::move(*m_deserialized_log_event)};
+        clear_deserialized_log_event();
+        return released;
+    }
+
+    /**
+     * Handles the incomplete IR error returned from `Deserializer::deserialize_next_ir_unit`.
+     * @param err
+     * @return true if the error is resolved on success.
+     * @return false with the relevant Python exception and error set on failures:
+     * - The error doesn't indicate an incomplete IR stream.
+     * - Incomplete IR stream is not allowed.
```

Comment:
Now I get what you mean, but I feel there might be better way to phrase it. Here is what I can think of, @kirkrodrigues can you give some input?
```suggestion
     * @return false with the relevant Python exception and error set on the following cases:
     * - The error doesn't indicate an incomplete IR stream.
     * - The error indicates an incomplete IR stream, and `allow_incomplete_stream` == `false`.
```

Reply:
I have some suggestions but the function's name is a bit confusing. If the error from the deserializer *wasn't* an incomplete IR error, then why is this method trying to handle it? Can we pick a name like `handle_deserializer_error` or `handle_deserialization_error`?

Reply:
This function is supposed to only handle incomplete IR errors. However, we have to check the error code to see if it's an incomplete error. So do u think it makes more sense to either:
- Rename this function to `handle_deserializer_error` and call incomplete IR error handler inside
- Check the error code before calling this function, and only call it when the error code indicates incomplete IR error

Notice that at this moment we don't handle any other errors except incomplete IR error. For any other errors we just need to return by setting the Python exception

Reply:
> Rename this function to handle_deserializer_error and call incomplete IR error handler inside

If you're renaming `handle_incomplete_ir_error` to `handle_deserializer_error`, then what do you mean by "call incomplete IR error handler inside"?

> Notice that at this moment we don't handle any other errors except incomplete IR error. For any other errors we just need to return by setting the Python exception

I would argue that "setting the Python exception" *is* handling the error. When we say we've handled all errors, it doesn't mean that every error is resolvable. It just means that this layer of the software does something for every error instead of crashing or invoking undefined behaviour. The docstring is what clarifies the behaviour for each error.

Reply:
hmmm, from the code behaviour I think we should check the error code before calling this function. So far we are only trying to resolve incomplete IR errors. For the rest of errors we should just throw them to users.
But I don't agree on the point that "setting the Python exception" is "handling errors." It's just like u're programming in only C++ and u throw whatever exception occurred, and we're arguing "throwing" == "handling" (the only difference is our exception cross ffi layer). Or maybe we can be more clear rather than using the word "handle". Based on the discussion, do u have better suggestion?

Reply:
If "handle" doesn't imply doing something with the error, then what is the difference between "resolve" and "handle"? Let's say we wanted to set a Python exception for every error, what would you call that method?

Reply:
- This function is not just set a Python exception for the given error. For incomplete IR error, we might just swallow it if users don't care
- If the "handler" is just to set an error, I would probably call it `set_python_exception(error_code)`, or as chatgpt suggested `map_error_to_exception`

Reply:
> This function is not just set a Python exception for the given error. For incomplete IR error, we might just swallow it if users don't care

Doesn't really answer my question of what is the difference between "handle" and "resolve".

Anyway, it's probably best you check the error outside the method.